sql: do not recurse the initialization of local execution via Start()

knz opened the pull request:
As stated earlier I want to clamp down on the amount of knowledge about the structure of the logical plan that's needed to perform local execution. In particular, eventually, the execution methods should not have access to the logical plan directly -- only the execution structure -- and thus should not need / be able to fish the "children" from the logical plan.

This first PR make a step in that direction, by extracting the recursion for the initialization step (Start) from the implementation of planNode.Start() methods into the previously-defined walkPlan(). The remaining initialization code stays in a new method called `startExec`, only implemented for nodes that need to initialize something. The result of this is that the remaining startExec() methods do not need to know about the children nodes, only their own nodes.

If we can make this happen successfully, I'll follow the same pattern in a subsequent PR for the part of Close() that pertains to local execution.  After these two have been segregated, I'll work on Next/Values which are comparatively much more easy.

@jordanlewis @RaduBerinde this code as it stands uses a closure to define the `planObserver`. Do you think this is allocation-heavy? If so, I can turn the planObserver into a proper visitor.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20416)
<!-- Reviewable:end -->

knz committed:
sql: make the planObserver interface smarter

knz committed:
sql: remove the planNode Start() method

The Start() method on planNode had two roles:
- initialize local execution
- recurse into sub-plans.

Eventually we want the code for local execution to become independent
from the logical planning, and in particular it should not directly
know/assume how to access the "children nodes" of a given relational
operator. In order to make the code agnostic, we can delegate the task
of the recursion to a suitable dispatch method.

It so happens that we already have such a dispatch function -
walkPlan() already used for other things.

This patch thus isolates the code that initializes local execution in
a new method startExec() which doesn't need to recurse, and changes
startPlan() to call this method via walkPlan().

knz committed:
sql: make startExec() start sub-queries itself

knz committed:
sql: simplify startPlan()

.. and also simplify the SCRUB implementation to use the `runParams`
struct more.

jordanlewis reviewed (approved):
LGTM!
<!-- review by re -->

RaduBerinde commented:
LGTM, good stuff!

knz commented:
TFYRs!
